movement/filtering.py

```diff
@@ -0,0 +1,138 @@
+import logging
+from copy import copy
+from datetime import datetime
+from functools import wraps
+from typing import Union
+
+import numpy as np
+import xarray as xr
+
+
+def log_to_attrs(func):
+    """
+    Appends log of the operation performed to xarray.Dataset attributes
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        result = func(*args, **kwargs)
+
+        log_entry = {
+            "operation": func.__name__,
+            "datetime": str(datetime.now()),
+            **{f"arg_{i}": arg for i, arg in enumerate(args[1:], start=1)},
+            **kwargs,
+        }
+
+        # Append the log entry to the result's attributes
+        if result is not None and hasattr(result, "attrs"):
+            if "log" not in result.attrs.keys():
+                result.attrs["log"] = []
+            result.attrs["log"].append(log_entry)
+
+        return result
+
+    return wrapper
+
+
+def filter_diagnostics(ds: xr.Dataset):
+    """
+    Reports the number of datapoints filtered
+    """
+
+    # Compile diagnostic report
+    diagnostic_report = "\nDatapoints Filtered:\n"
+    for ind in ds.individuals.values:
+        diagnostic_report += f"\nIndividual: {ind}"
+        for kp in ds.keypoints.values:
+            n_nans = np.count_nonzero(
+                np.isnan(
+                    ds.pose_tracks.sel(individuals=ind, keypoints=kp).values[
+                        :, 0
+                    ]
+                )
+            )
+            n_points = ds.time.values.shape[0]
+            prop_nans = round((n_nans / n_points) * 100, 1)
+            diagnostic_report += (
+                f"\n   {kp}: {n_nans}/{n_points} ({prop_nans}%)"
+            )
+
+    # Write diagnostic report to logger
+    logger = logging.getLogger(__name__)
+    logger.info(diagnostic_report)
```

Comment:
I'm currently struggling a bit to see how this approach is more convenient for the user than simply printing the diagnostics, as we did previously. However, I admit that this may just be because I'm unfamiliar with Python logging! What is the expected interface here? How exactly is the user meant to access these logs when they want to? Would it really not be easier to simply allow user to toggle printing of the diagnostics on and off?

Reply:
This is natural question to ask if you haven't worked with logging before (as I myself hadn't a year ago). Actually logging is made to handle both the writing to log files and printing to the terminal (this can be configured).
The way I understand it, logging has several advantages to just printing:
- there are different [logging levels](https://docs.python.org/3/library/logging.html#logging-levels), ranging in severity. These means that messages can be "annotated" so that others know whether sth is e.g. a warning or just some useful info.
- logging can handle both printing to terminal (like "print") as well as writing to text files. Moreover, what it prints is not just the message, but also relevant context information about it, e.g datetime and which code line trigerred the error.
  ```console
  2023-04-28 12:32:09,506 - ERROR - MainProcess validators.py:29 - File not found: .data/DLC_sample_poses.csv
  ```
  What information is included in the log can be configured by us. Needless to say this is very valuable for the users and for us during debugging.
- the different logging levels allow us to configure handling logic for them. For example, we can choose to log all levels to the log file (incl. DEBUG) and only show INFO, WARNING, and ERROR in the terminal. This means that we can still keep lots of granular information useful for debugging, without bombarding the user with countless messages.

I hope this is clearer now. FYI you can find the movement logs in `~/.movement/movement.log`. Logs generated during testing are sent to a different, temporary location.